### PR TITLE
[Go] Add support for Windows Golang tests

### PR DESCRIPTION
**Pull request by erezgeva**
This PR is based on three PRs

1. <del> #3355 </del>
2. <del> #3357  </del>
3. <del> #3349 </del>

This PR commit:
Add Windows support of Golang tests

- Add 2 variant of Golang tests:
  - MingW64
  - MingW64 with UCRT
- Use `cygpath` to convert include pathes to windows format.
  Golang on windows, expect windows format.
- Use `BOOST_CPPFLAGS` directly, as we need to convert it using the `cygpath` tool.
- Improve Golang test syntax.
- Support building on same folder of source code,
  i.e. `SRCDIR` can be empty!
- Use `mkdir` with `-p`, it is supported by POSIX for more than 20 years,
  https://pubs.opengroup.org/onlinepubs/009695399/utilities/mkdir.html
 after all our minimum Golang version 1.20 is from 2023, https://go.dev/doc/devel/release.
  Remove the error avoidance `|| true` and error output dump `2>/dev/null1`.
  In case of error, we need to stop and find the proper location of the error!
  The use of `-p` will avoid the existing folder warning and
  also create all the parents folder needed.
- Use the `tr` to clean the multiple cpp test list records
  before calling the compiler helper.
- `ROOT_DIR` is NOT a Javascript parameter,
  move up to the main parameters list.
- Add optional debug flag `GOBUILDDBG` for `go build`,
  So we can add the `-x` flag to print actual commands.
- We need GNU extension to build Cgo windows GCC.
  Use `-std=gnu11` or newer.
  See used source code compiled for `cgo`:
  https://github.com/golang/go/tree/master/src/runtime/cgo/gcc_windows_amd64.c

Note for @wsfulton ,
Tests were added for testing new features and new ability to test.
Fill free to reduce the number of tests, and you may increase there diversity, using the available matrix variables :smile: 
Be ware that not all combinations works.
For example, go require the C gnu extension standards for the `cgo` source code compilation (yes, part of `go` deployment on windows).

### CODE REVIEW

File: Examples/test-suite/go/Makefile.in [hunk 1]
```diff
@@ -152,21 +180,20 @@ run_multi_testcase = \
...
+	      imports_a.go imports_a.gox imports_b.go imports_b.gox \
+	      mod_a.go mod_a.gox mod_b.go mod_b.gox \
+	      multi_import_a.go multi_import_a.gox \
+	      multi_import_b.go multi_import_b.gox \
+	      multi_import_d.go multi_import_d.gox \
+	      packageoption_a.go packageoption_a.gox \
+	      packageoption_b.go packageoption_b.gox \
+	      packageoption_c.go packageoption_c.gox \
+	      template_typedef_cplx2.go template_typedef_cplx2.gox
+	rm -rf go_subdir_import_a.go go_subdir_import_a.gox testdir gopath
```
Comment (wsfulton) [inline Examples/test-suite/go/Makefile.in]: What's the point of this change? \ is unpleasant on the eyes, so prefer no continuation lines please.

File: Examples/test-suite/go/Makefile.in [hunk 2]
```diff
@@ -62,9 +66,8 @@ li_windows.cpptest:
 
 go_subdir_import.multicpptest:
 	$(setup)
-	$(go_multicpp_setup)
-	mkdir -p testdir/go_subdir_import 2>/dev/null || true
-	mkdir -p gopath/go_subdir_import/src/testdir/go_subdir_import 2>/dev/null || true
+	mkdir -p testdir/go_subdir_import \
+	         gopath/go_subdir_import/src/testdir/go_subdir_import
```
Comment (wsfulton) [inline Examples/test-suite/go/Makefile.in]: Why break this line?

File: .github/workflows/windows.yml [hunk 3]
```diff
@@ -265,8 +337,14 @@ jobs:
...
-            cl.exe /? 2>&1 | head -n1
+            cl.exe -? 2>&1 | head -n1
           else
+            # TODO: check if mingw-w64 'isfinite()' is updated
+            # We wait for '__fpclassify(x),' => '__fpclassify((double)(x)),
+            grep -A3 '#define fpclassify(x)' '/c/mingw64/x86_64-w64-mingw32/include/math.h' || true
+            if [[ "$MSYSTEM" = 'UCRT64' ]]; then
+              grep -A3 '#define fpclassify(x)' '/ucrt64/include/math.h' || true
+            fi
             which gcc
```
Comment (wsfulton) [inline .github/workflows/windows.yml]: How is this related to "add support for Windows Golang tests"? I think if something is broken, then don't test for it. We are only testing for SWIG problems, and if something is broken in the system preventing that then users are not likely to be using SWIG. We need to keep these tests simple and easily maintainable.

**Comment by wsfulton** [comment]
@erezgeva, we need to keep the testing to a bare minimum instead of testing so many versions of Go; it takes too long to test so many combinations of platforms, c++ versions, compiler versions, target language versions etc. The approach we have taken and need to keep to is more comprehensive testing on linux and more sanity bare minimum testing on the more expensive proprietary OSs (windows/MacOS). So on the latter, just test one or two versions of target languages max. For windows and MinGW/Cygwin/msys, just test one of each really and if they have problems, then just don't test on them.

**Comment by erezgeva** [comment]
> need to keep the testing to a bare minimum instead of testing so many versions of Go; it takes too long to test so many combinations of platforms, c++ versions, compiler versions, target language versions etc. The approach we have taken and need to keep to is more comprehensive testing on linux and more sanity bare minimum testing on the more expensive proprietary OSs (windows/MacO

As i note above:
> Tests were added for testing new features and new ability to test.
> Fill free to reduce the number of tests

I will leave 2 tests with new `Go`.
The install go package from MSYS2 is version 1.26

File: .github/workflows/windows.yml (hunk 3 above)
Comment (erezgeva) [reply .github/workflows/windows.yml]: This comes from:
- #3349

As I know that the `Go` test pass without warnings.
I will remove the patch from this PR.

File: Examples/test-suite/go/Makefile.in (hunk 2 above)
Comment (erezgeva) [reply Examples/test-suite/go/Makefile.in]: I thought it is too long,
I remove the line break.

File: Examples/test-suite/go/Makefile.in (hunk 1 above)
Comment (erezgeva) [reply Examples/test-suite/go/Makefile.in]: These are 16 separate `rm -f` .
I thought of speeding up.

I revert.
